crates/sui-verifier/src/char_type_verifier.rs

```diff
@@ -0,0 +1,227 @@
+                .map_err(verification_failure)?;
+            // if we reached this point, it means we have a legitimate characteristic type candidate
+            // and we have to make sure that both the init function's signature reflects this and
+            // that this type is not instantiated in any function of the module
+            char_type_candidate = Some((struct_name, struct_handle, def));
+            break; // no reason to look any further
+        }
+    }
+    for fn_def in &module.function_defs {
+        let fn_handle = module.function_handle_at(fn_def.function);
+        let fn_name = module.identifier_at(fn_handle.name);
+        if fn_name == INIT_FN_NAME {
+            if let Some((struct_name, struct_handle, _)) = char_type_candidate {
+                // only verify if init function conforms to characteristic types requirements if we
+                // have a characteristic type candidate
+                verify_init_function_char_type(module, fn_handle, struct_name, struct_handle)
+                    .map_err(verification_failure)?;
+            } else {
+                // if there is no characteristic type candidate than the init function should have
+                // only one parameter of TxContext type
+                verify_init_function_single_param(module, fn_handle)
+                    .map_err(verification_failure)?;
+            }
+        }
+        if let Some((struct_name, _, def)) = char_type_candidate {
+            // only verify lack of characteristic types instantiations if we have a
+            // characteristic type candidate
+            verify_no_instantiations(module, fn_def, struct_name, def)
+                .map_err(verification_failure)?;
+        }
+    }
+
+    Ok(())
+}
+
+// Verifies all required properties of a characteristic type candidate (that is a type whose name is
+// the same as the name of a
+fn verify_char_type(
+    module: &CompiledModule,
+    struct_name: &str,
+    struct_handle: &StructHandle,
+    struct_def: &StructDefinition,
+) -> Result<(), String> {
+    // must have only one ability: drop
+    let drop_set = AbilitySet::from_u8(Ability::Drop as u8).unwrap();
+    let abilities = struct_handle.abilities;
+    if !(abilities.has_drop() && (abilities.union(drop_set)) == drop_set) {
+        return Err(format!(
+            "characteristic type candidate {}::{} must have a single ability: drop",
+            module.self_id(),
+            struct_name,
+        ));
+    }
+    let field_count = struct_def.declared_field_count().map_err(|_| {
+        format!(
+            "characteristic type candidate {}::{} cannot be a native structure",
+            module.self_id(),
+            struct_name
+        )
+    })?;
+
+    // unwrap below is safe as it will always be successful if declared_field_count call above is
+    // successful
+    if field_count != 1
+        || struct_def.field(0).unwrap().signature != TypeSignature(SignatureToken::Bool)
```

Comment:
`signature.0` to avoid the `TypeSignature(...)` wrapper

Reply:
You explicitly tried to avoid it as if I remember correctly you don't love `.0` syntax :-) Changed now!

Reply:
I'm here and there, maybe this is going back to my love for my little `sp!()` macro in the source language? I do tend to dislike field access instead of pattern matching